File: src/main/kotlin/com/full/stack/assignment/f1/data/cache/mapper/CircuitMapper.kt
package com.full.stack.assignment.f1.data.cache.mapper

import com.full.stack.assignment.f1.data.cache.entity.CircuitEntity
import com.full.stack.assignment.f1.data.cache.entity.CircuitLocationEntity
import com.full.stack.assignment.f1.model.Circuit
import com.full.stack.assignment.f1.model.CircuitLocation
import org.springframework.stereotype.Component

@Component
class CircuitMapper: Mapper<Circuit, CircuitEntity> {
    override fun toDomain(entity: CircuitEntity): Circuit {
        return Circuit(
            id = entity.id,
            name = entity.name,
            location = CircuitLocation(
                locality = entity.location.locality,
                country = entity.location.country,
            ),
        )
    }

    override fun toEntity(domain: Circuit): CircuitEntity {
        return CircuitEntity(
            id = domain.id,
            name = domain.name,
            location = CircuitLocationEntity(
                locality = domain.location.locality,
                country = domain.location.country,
            )
        )
    }
}

File: frontend/lib/data/remote/mappers/circuit_mapper.dart
import 'package:f1_api_client/api.dart';
import '../../../domain/models/circuit/circuit.dart';
import '../../../helpers/mapper/mapper.dart';

class CircuitMapper extends Mapper<CircuitDTO, Circuit, void> {
  @override
  Circuit map(CircuitDTO param, {void metadata}) {
    return Circuit(
      circuitId: param.circuitId,
      circuitName: param.circuitName,
      locality: param.location.locality,
      country: param.location.country,
    );
  }
}
